src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.java

```diff
@@ -235,27 +226,105 @@ private void turnOffReadiness(Throwable exception){
             throw new IllegalStateException("Failed to create RabbitMQ connection using configuration", e);
         }
 
-        this.connection.addBlockedListener(new BlockedListener() {
-            @Override
-            public void handleBlocked(String reason) throws IOException {
-                LOGGER.warn("RabbitMQ blocked connection: {}", reason);
+        addBlockedListenersToConnection(this.connection);
+        addRecoveryListenerToConnection(this.connection);
+    }
+
+    private void addShutdownListenerToChannel(Channel channel, Boolean withRecovery) {
+        channel.addShutdownListener(cause -> {
+            LOGGER.debug("Closing the channel: ", cause);
+            if (!cause.isHardError() && cause.getReference() instanceof Channel) {
+                Channel channelCause = (Channel) cause.getReference();
+                Method reason = cause.getReason();
+                if (reason instanceof AMQImpl.Channel.Close) {
+                    var castedReason = (AMQImpl.Channel.Close) reason;
+                    if (castedReason.getReplyCode() != 200) {
+                        StringBuilder errorBuilder = new StringBuilder("RabbitMQ soft error occupied: ");
+                        castedReason.appendArgumentDebugStringTo(errorBuilder);
+                        errorBuilder.append(" on channel ");
+                        errorBuilder.append(channelCause);
+                        String errorString = errorBuilder.toString();
+                        LOGGER.warn(errorString);
+                        if (withRecovery && errorString.contains("PRECONDITION_FAILED")) {
+                            recoverSubscriptionsOfChannel(channel);
+                        }
+                    }
+                }
             }
+        });
+    }
 
-            @Override
-            public void handleUnblocked() throws IOException {
-                LOGGER.warn("RabbitMQ unblocked connection");
+    private void recoverSubscriptionsOfChannel(Channel channel) {
+        channelChecker.execute(() -> {
+            var mapEntries =
+                    channelsByPin
+                            .entrySet()
+                            .stream()
+                            .filter(entry -> Objects.nonNull(entry.getValue().channel) && channel.getChannelNumber() == entry.getValue().channel.getChannelNumber())
+                            .collect(Collectors.toList());
+            for (Map.Entry<PinId, ChannelHolder> mapEntry : mapEntries) {
```

Comment:
We should only one channel because manager create a separate channel for each pin